posts/2024-06-05-mixing-python-envs.md

```diff
@@ -0,0 +1,181 @@
+    <terminal-line data="input">python3 -m venv xwmb_venv --system-site-packages</terminal-line>
+    <terminal-line data="input">source xwmb_venv/bin/activate</terminal-line>
+    <terminal-line data="input" inputChar="(xwmb_venv) $">pip install git+https://github.com/hdrake/xwmb.git@main</terminal-line>
+</terminal-window>
+
+In a similar vein, if you're installing more packages into a virtual environment and they
+begin to conflict with each other, the best course of action is to create a new virtual environment.
+Virtual environments are isolated from each other, so rather than creating large, monolithic 
+environments, it is common practice to create virtual environments for a single package, and have
+multiple virtual environments depending on your immediate requirements.
+
+In general, if anything in the environment doesn't work for whatever reason, the best course of action
+is to delete it and try again. You may need to try different `conda/analysis3` environments, or, if none of those
+work, use one of the NCI installed Python modules as your base. 
+```{note}
+NCI's Python modules do not have `dask`, `xarray`, etc. installed, so the resultant virtual environments could be quite large.
+```
+
+
+## Jupyter Kernel
+There is one more step necessary to have your new virtual environment usable as a kernel in an ARE JupyterLab session. 
+You'll need to call `ipykernel install` while the virtual environment is activated.
+<terminal-window typingDelay=30 lineDelay=200>
+    <terminal-line data="input">module use /g/data/hh5/public/modules</terminal-line>
+    <terminal-line data="input">module load conda/analysis3</terminal-line>
+    <terminal-line data="input">source xwmb_venv/bin/activate</terminal-line>
+    <terminal-line data="input" inputChar="(xwmb_venv) $">python3 -m ipykernel install -\\\-user -\\\-name xwmb-venv -\\\-display-name "xWMB + conda/analysis3"</terminal-line>
+</terminal-window>
+
+```{note}
+Since this virtual environment is built on top of `conda/analysis3`, you do not need to install `ipykernel` in it, as this is already present in the conda environment. 
+```
+
+Now, when
+you launch a new ARE JupyterLab instance, you will be able to see your new environment,
+launch a notebook and import the `xWMB` package
+
+![Jupyterlab launch](../images/notebook_virtualenv_launch.gif)
+
+You only have to do this once. If you delete the environment and recreate it at the same path,
+Jupyterlab will still be able to use it.
+
+## Summary
+
+While CMS tries to make the `hh5` conda analysis environments as comprehensive as possible, we can't
+always install the packages you need. By creating a virtual environment using the `conda/analysis3` 
+environment as your base,
+its possible to add in packages CMS can't install, without having the overhead of maintaining
+your own large conda environment. Furthermore, using a virtual environment avoids the 
+headaches associated with installing software using `pip install --user` as anything 
```

Comment:
remove "avoids headaches"  just say it's more stable and easier to maintain and reuse that ....